### ISSUE TITLE
Address Fall 2021 TCCNS Updates

### ISSUE THREAD

**Issue by au5ton** [cougargrades/publicdata#9]
Referenced here: https://uh.edu/academics/courses-enrollment/course-number-updates/index.php

> In an effort to make transferring credit for students easier, several University of Houston course names and numbers are changing to align with the TCCNS. These changes go into effect starting in Fall 2021.
> Below is a chart outlining all course number changes, organized alphabetically by current UH course number. For more information related to these course changes please see the FAQ page.
> The Texas Common Course Numbering System (TCCNS) is a cooperative effort among Texas community colleges and universities to facilitate transfer of freshman- and sophomore-level general academic courses.

I'm curious how this can be addressed in the cougargrades dataset.

Some initial thoughts:
- Is this even a problem? Can't we just leave them be and start a new "course"?
- If we were to implement some kind of new "reference" system, what would be the best way to do this?
- Is it possible for a new course number to conflict with an old course number for a different course?
    - Example: 
    - "calc 1 (math 1500)"  and "geometry (math 1300)" are courses.
    - "geometry" was discontinued under the course number "math 1300" many years ago, but still exists in old sections
    - "calc 1" has its course number changed to "math 1300"
    - the course number "math 1300" now means 2 completely different courses across a large time period

**Comment by au5ton** [comment]
A potential solution for this would be to:
- Archive all the course number changes referenced in the link above
- Add a "notices[]" field to the `Course` type
    - This `notice` would have 3 properties:
        - `title`: string, headline which introduces the notice
        - `emoji`: string, icon which denotes the "type" of notice without needing to keep a separate enum and UI up-to-date
        - `body`: string, content which explains the notice or delivers the information (maybe markdown? i don't want to wastefully add dependencies, but I want this to be useful)
        - `href`:  string | undefined, URL where the full explanation or referenced material can be found
        - `publish`: string | undefined, ISO timestamp when this notice should start being rendered
        - `expiry`: string | undefined, ISO timestamp when this notice should stop being rendered
- Write a Patchfile which will provide a notice to both the old course number and the new course number where they would be rendered on the course page

**Comment by au5ton** [comment]
An alternative solution for this would be:
- add a "relatedCourses[]" field to all courses which are affected by these changes
- in this field, document references are added to their corresponding counterpart courses
  - `previousCourse.relatedCourses = [ currentCourse ]`
  - `currentCourse.relatedCourses = [ previousCourse ]`
- this structure of information can be used to allow the web interface to dynamically supplement the information is loads with information from the previous course
- perhaps 2 fields like this can be added to indicate history, such as `pastCourseNumbers[]` and `futureCourseNumbers[]` so that the UI can say something like "previously known as COOG 1234" or "replaced with COOG 1234"

**Comment by au5ton** [comment]
Perhaps, to convey _when_ a course number has changed, a full object could be specified:

```typescript
/**
 * UI message:
 *  on the new course: "Previously known as COOG 1234 (effective Fall 2021)."
 *  on the old course: "Replaced with COOG 1234 (effective Fall 2021)."
 */

course.tccnsUpdates = [
  {
    effective: 202103, // Fall 2021
    relation: 'past', // 'past' | 'future'
    courseHref: '/c/COOG 1234', 
    sourceHref: 'https://uh.edu/...'
  }
]
```

**Comment by au5ton** [comment]
Revision:

```typescript
course.tccnsUpdates = [
  {
    shortMessage: 'Renamed to COOG 1111' | 'Previously known as COOG 1234',
    longMessage: 'Effective Fall 2021, COOG 1234 was renamed to COOG 1111',
    courseHref: '/c/COOG 1234', 
    sourceHref: 'https://uh.edu/...'
  }
]
```

**Comment by au5ton** [comment]
Things to consider:
- `tccnsNew2Old` patchfiles can't be used until Fall 2021 data has been acquired because they make changes to courses which don't "exist" yet
- once Fall 2021 data has been acquired, `publicationlink` patchfiles and sitemap.txt will generate correctly
- core curriculum data with new course numbers has already been added and [patchfiles don't conditionally generate](https://github.com/cougargrades/publicdata/blob/579b47a5ab099f45bce09e16900528c3f3c0a169/bundler/bundle/patch/publications_core.py#L12-L31)
- basically, integration can't complete until: a) Fall 2021 data has been acquired, b) a variety of changes are made to include course "stubs" (empty courses), which is made complicated by some patchfile generation requiring that CSV data exists

tl;dr database regen required once Fall 2021 data is added
